[Deprecated] Starting AI/ML section in python docs

xhlulu opened the pull request:
Please refer to: https://github.com/plotly/plotly.py/issues/2153

Changes:
- Added scikit-learn to requirements

Todo list:
- [x] Create a ML section
- [x] kNN Classification notebook
- [x] Regression notebook
- [ ] ROC notebook
- [ ] PCA notebook
- [ ] tSNE/UMAP notebook
- [ ] LIME (follow [this example](https://marcotcr.github.io/lime/tutorials/Tutorial%20-%20images.html))

emmanuelle commented:
Hi @xhlulu  thanks for the pull request :-). Doc tutorials need to have a header with information from jupytext and some meta-data about the tutorial. See for example the header of https://raw.githubusercontent.com/plotly/plotly.py/master/doc/python/line-and-scatter.md. I'm surprised that there is no header at all did you edit the file in the notebook? jupyter/jupytext should have created one. 
For the header you can copy&paste from another md file, and edit the information in a text editor (like the title etc.).  In which section of the doc should this tutorial appear? A new one maybe?

emmanuelle commented on doc/python/ml-knn.md:
```diff
@@ -0,0 +1,119 @@
+X, y = make_moons(noise=0.3, random_state=0)
+
+# Create a mesh grid on which we will run our model
+x_min, x_max = X[:, 0].min() - margin, X[:, 0].max() + margin
```
margin is not defined here

emmanuelle commented on doc/python/ml-knn.md:
```diff
@@ -0,0 +1,119 @@
+        colorscale=[[0.0, 'blue'], [0.5, 'red'], [1.0, 'green']],
+        opacity=0.25
+    )
+)
```
Nice example; in order to show some added value of plotly you could compute also the probability of each label and display it in the hover info, like below
```
import numpy as np
from sklearn.neighbors import KNeighborsClassifier
import plotly.express as px
import plotly.graph_objects as go
mesh_size = .02
margin = 1
df = px.data.iris()
X = df[['sepal_length', 'sepal_width']]
y = df.species_id
# Create a mesh grid on which we will run our model
l_min, l_max = df.sepal_length.min() - margin, df.sepal_length.max() + margin
w_min, w_max = df.sepal_width.min() - margin, df.sepal_width.max() + margin
lrange = np.arange(l_min, l_max, mesh_size)
wrange = np.arange(w_min, w_max, mesh_size)
ll, ww = np.meshgrid(lrange, wrange)
# Create classifier, run predictions on grid
clf = KNeighborsClassifier(15, weights='distance')
clf.fit(X, y)
Z = clf.predict(np.c_[ll.ravel(), ww.ravel()])
Z = Z.reshape(ll.shape)
proba = clf.predict_proba(np.c_[ll.ravel(), ww.ravel()])
proba = proba.reshape(ll.shape + (3,))
fig = px.scatter(df, x='sepal_length', y='sepal_width', color='species')
fig.update_traces(marker_size=10, marker_line_width=1)
fig.add_trace(
    go.Heatmap(
        x=lrange, 
        y=wrange, 
        z=Z, 
        showscale=False,
        colorscale=[[0.0, 'blue'], [0.5, 'red'], [1.0, 'green']],
        opacity=0.25,
        customdata=proba,
        hovertemplate='sepal length: %{x} <br>sepal width: %{y} <br>p(setosa): %{customdata[0]:.3f}<br>p(versicolor): %{customdata[1]:.3f}<br>p(virginica)%{customdata[2]:.3f} <extra></extra>'
    )
)
```

emmanuelle commented on doc/python/ml-knn.md:
````diff
@@ -0,0 +1,119 @@
+        opacity=0.25
+    )
+)
+```
````
Please add `fig.show` at the end of each example, this is part of our stylistic guidelines (which we should document!)

xhlulu commented:
Thanks @emmanuelle for the review! I added your suggestions to the notebook. Here's an overview of what I also changed: _"i split the knn into two notebooks; one for regression (didn't touch it further, just a placeholder notebook), and added 4 sections to the knn notebook. Now it has two "simple" sections that takes less than 20 lines, and 2 more "advanced vis" for 3D and splom viz"_

xhlulu commented:
Some things I am stuck on:
- In the header, I need to indicate a `permalink`, and a `redirect_from`. Is the permalink the url of the notebook? As for the redirect from, is it the URL of the section where the notebook is categorized?
- Do I need to manually create a "ML section" in some other files before the new section appears, or is it done automatically when I create a md notebook?
- Where should I store my new thumbnail?
- How can you create a new notebook such that it has an automatic header generated, and is in md format? what i'm currently is manually create a new md file, and then close it and reopen it and it will load it as a notebook.

For the header, here's what it currently looks like: https://github.com/plotly/plotly.py/pull/2172/files/70f7131b939e4396b46405b02e42b3914c9bdfe7#

nicolaskruchten commented:
Good questions! @jdamiba shouls be able to steer you in the right direction :)

jdamiba commented:
> Some things I am stuck on:
> 
> * In the header, I need to indicate a `permalink`, and a `redirect_from`. Is the permalink the url of the notebook? As for the redirect from, is it the URL of the section where the notebook is categorized?
> * Do I need to manually create a "ML section" in some other files before the new section appears, or is it done automatically when I create a md notebook?
> * Where should I store my new thumbnail?
> * How can you create a new notebook such that it has an automatic header generated, and is in md format? what i'm currently is manually create a new md file, and then close it and reopen it and `jupyter` will load it as a notebook.
> 
> For the header, here's what it currently looks like: https://github.com/plotly/plotly.py/pull/2172/files/70f7131b939e4396b46405b02e42b3914c9bdfe7#

The `permalink` determines the URL of the notebook. This is always a [slugified](https://blog.tersmitten.nl/slugify/) version of the `name`, preceded by `/python/`.

So if the `name` of the notebook is `Example Title` then the `permalink` will be `/python/example-title`. If the `name` of the notebook is `Another Example Title` then the `permalink` will be `/python/another-example-title`.

The `redirect_from` is only necessary if you want the notebook to take over traffic from another notebook that is being deprecated. In this case you don't need to use it since you are creating new URLs. 

You don't need to create the ML section, I took care of that in https://github.com/plotly/graphing-library-docs/pull/9. Just make sure to set `display_as: ai_ml`. Posts that will go on the index page need to have `page_type: example_index` and `order` less than five. Posts with `order` greater than five need to have `page_type: u-guide`

Here are the instructions for uploading a thumbnail image. Once the image is uploaded, you set `thumbnail: thumbnail/your-image-name.jpg`.

```
Use images.plot.ly for adding new images. The password is in the Plotly 1Password Engineering Vault.
Log-in here: https://661924842005.signin.aws.amazon.com/console
From the Amazon Web Services Console select S3 (Scalable Storage in the Cloud) then select plotly-tutorials -> plotly-documentation -> thumbnail
Now from All Buckets /plotly-tutorials/plotly-documentation/thumbnail select the Actions dropdown and upload your .jpg file
```

I'm not sure how to create a new notebook with the right header- I would suggest just copy and pasting the header from another example and modifying it. Maybe @emmanuelle has a better idea? You can also try creating the notebook in Jupyter if you get the Jupytext extension. 

Let me know if you have any other questions or if any of this wasn't clear!

xhlulu commented:
@jdamiba Wow! Thank you for the super detailed answer! I'll be working on this friday/saturday. Just a question: does `order` start at 1 or 0? And if i understand correctly, we would be doing `0,1,2,3,4` for what we want to show on the landing page, then `6,7,8,...` for everything else? I show never use `5`?

jdamiba commented:
Order actually starts at 1, so you want to use 1, 2, 3, and 4 for the pages you want to show up on the landing page, and 6, 7, 8, ..., N for the rest. 

No page that you create in this repo should have order 5, as that is reserved for use by the category index page in the `graphing-library-docs` repo.

emmanuelle commented on doc/python/ml-knn.md:
```diff
@@ -0,0 +1,251 @@
+from sklearn.neighbors import KNeighborsClassifier
+
+X, y = make_moons(noise=0.3, random_state=0)
+X_test, _ = make_moons(noise=0.3, random_state=1)
```
Here it would be more "scikit-learn-idiomatic" to use scikit-learn's [`train_test_split`](https://scikit-learn.org/stable/modules/generated/sklearn.model_selection.train_test_split.html#sklearn.model_selection.train_test_split) (and possibly to generate a larger dataset)

emmanuelle commented on doc/python/ml-knn.md:
```diff
@@ -0,0 +1,251 @@
+from sklearn.neighbors import KNeighborsClassifier
+
+X, y = make_classification(n_features=2, n_redundant=0, random_state=0)
+X_test, _ = make_classification(n_features=2, n_redundant=0, random_state=1)
```
Here also I would use `train_test_split`.

emmanuelle commented on doc/python/ml-knn.md:
```diff
@@ -0,0 +1,251 @@
+clf.fit(X, y)  # Fit on training set
+y_score = clf.predict_proba(X_test)[:, 1]  # Predict on new data
+
+fig = px.scatter(x=X_test[:, 0], y=X_test[:, 1], color=y_score, labels={'color': 'score'})
```
```suggestion
fig = px.scatter(x=X_test[:, 0], y=X_test[:, 1], color=y_score, 
                        labels={'color': 'score of <br>first class'})
```

emmanuelle commented on doc/python/ml-knn.md:
```diff
@@ -0,0 +1,251 @@
+proba = clf.predict_proba(np.c_[ll.ravel(), ww.ravel()])
+proba = proba.reshape(ll.shape + (3,))
+
+fig = px.scatter(df, x='sepal_length', y='sepal_width', color='species', width=1000, height=1000)
```
Here the figure is a bit too large, you have to scroll horizontally. 800 works better than 1000 I think. Also, if you imposed this to have a square figure, you can constrain the axis to have a square aspect ration, like
```
fig.update_layout(yaxis = dict(
      scaleanchor = "x",
      scaleratio = 1,
    ))
```

xhlulu replied:
Sorry this was for generating the display pics haha

emmanuelle commented on doc/python/ml-knn.md:
```diff
@@ -0,0 +1,251 @@
+
+## High Dimension Visualization with `px.scatter_matrix`
+
+If you need to visualize classifications that go beyond 3D, you can use the [scatter plot matrix](https://plot.ly/python/splom/).
```
Please mention that selections in one subplot will cross-filters the other subplots, this is a really cool feature of splom.

xhlulu replied:
I ended up removing this plot :/

emmanuelle commented on doc/python/ml-knn.md:
```diff
@@ -0,0 +1,251 @@
+clf.fit(X_train, y_train)
+y_pred = clf.predict(X_test)
+
+fig = px.scatter_matrix(X_test, dimensions=features, color=y_pred, labels={'color': 'prediction'})
```
Maybe add `y_test` in the hover data?

emmanuelle commented on doc/python/ml-regression.md:
```diff
@@ -0,0 +1,474 @@
+### Ordinary Least Square (OLS) with `plotly.express`
+
+
+This example shows how to use `plotly.express` to train a simply Ordinary Least Square (OLS) that can predict the tips servers will receive based on the value of the total bill.
```
```suggestion
This example shows how to use `plotly.express` to train a simply Ordinary Least Square (OLS) that can predict the tips waiters will receive based on the value of the total bill.
```

emmanuelle commented on doc/python/ml-regression.md:
```diff
@@ -0,0 +1,474 @@
+from sklearn.linear_model import LinearRegression
+
+df = px.data.tips()
+X = df.total_bill.values.reshape(-1, 1)
```
`df.total_bill[:, None]` can be a bit more compact.

emmanuelle commented on doc/python/ml-regression.md:
```diff
@@ -0,0 +1,474 @@
+
+df = px.data.tips()
+X = df.total_bill.values.reshape(-1, 1)
+X_train, X_test, y_train, y_test = train_test_split(X, df.tip, random_state=0)
```
The testing set is not used for any testing here so I'm not sure I understand the goal of this example.

emmanuelle commented on doc/python/ml-regression.md:
```diff
@@ -0,0 +1,474 @@
+y = df['petal_width']
+
+# Condition the model on sepal width and length, predict the petal width
+knn = KNeighborsRegressor(10, weights='distance')
```
There is a lot of overfit with this classifier, you can see all the spikes trying to fit specific points. I'd rather use a `sklearn.svm.SVR` here (I checked it creates a much smoother surface). You can also make a figure with subplots showing one subplot with the `KNeighborsRegressor` (a lot of overfit) and the other one with the `SVR` (much less overfit).

xhlulu replied:
Done

emmanuelle commented on doc/python/ml-regression.md:
```diff
@@ -0,0 +1,474 @@
+x_range = np.linspace(X.min(), X.max(), 100).reshape(-1, 1)
+
+fig = px.scatter(df, x='total_bill', y='tip', opacity=0.65)
+for n_features in [1, 2, 3, 4]:
```
I would use `degree` instead of `n_features`, the degree of the polynom is different from the number of features.

emmanuelle commented on doc/python/ml-regression.md:
```diff
@@ -0,0 +1,474 @@
+model.fit(X, y)
+y_pred = model.predict(X)
+
+fig = px.scatter(x=y_pred, y=y, labels={'x': 'prediction', 'y': 'actual'})
```
```suggestion
fig = px.scatter(x=y, y=y_pred, labels={'x': 'ground truth', 'y': 'prediction'})
```

emmanuelle commented on doc/python/ml-regression.md:
```diff
@@ -0,0 +1,474 @@
+fig.add_shape(
+    type="line", line=dict(dash='dash'),
+    x0=y.min(), y0=y.min(), 
+    x1=y.max(), y1=y.max()
```
```suggestion
    x0=0, y0=0, 
    x1=y.max(), y1=y.max()
```

emmanuelle replied:
this is a very minor aesthetic point, but it looks nicer I think when going through `(0, 0)`

xhlulu replied:
I wanted to make the example as generic as possible, since the user might want to use a y value that could be negative as well.

emmanuelle commented on doc/python/ml-regression.md:
```diff
@@ -0,0 +1,474 @@
+model.fit(X_train, y_train)
+df['prediction'] = model.predict(X)
+
+fig = px.scatter(
```
Here I'd rather plot the prediction against the ground truth, ie `x='petal_width` and `y=prediction`.

emmanuelle commented on doc/python/ml-regression.md:
```diff
@@ -0,0 +1,474 @@
+    df, x='prediction', y='petal_width',
+    marginal_x='histogram', marginal_y='histogram',
+    color='split', trendline='ols'
+)
```
You can add here
```
fig.update_traces(histnorm='probability', selector={'type':'histogram'})
```
so that the histogram of ground truth and prediction are normalized in the same way, if you display the counts you have less counts for the testing set as it is smaller.

xhlulu replied:
didn't know about that, thanks!

emmanuelle commented on doc/python/ml-regression.md:
```diff
@@ -0,0 +1,474 @@
+y = df['lifeExp']
+
+# Train model to predict life expectancy
+model = LassoCV(cv=N_FOLD, normalize=True)
```
here `N_FOLD` is not defined, and the CI fails (to avoid this, you can do "restart and run all" on the whole notebook before pushing a commit).

xhlulu commented:
@emmanuelle thank you for all the suggestions! I just want to first draft out the other notebooks, then address your reviews :)

nicolaskruchten commented:
Moved to https://github.com/plotly/plotly.py/pull/2276
